dcos_launch/terraform.py

```diff
@@ -0,0 +1,285 @@
+                               stderr=subprocess.STDOUT)
+            except FileNotFoundError:
+                log.info('No Terraform installation detected. Terraform is now being installed.')
+                self._install_terraform()
+
+            if self.config['key_helper']:
+                self.key_helper()
+            else:
+                log.warning('WARNING: {}Since you did not set "key_helper: true" in your config, make sure your '
+                            'ssh-agent is running i.e. "eval `ssh-agent -s`" and that you have added your private key '
+                            'to it i.e. "ssh-add /path/to/key.pem". ssh-agent usage is specific to terraform, not '
+                            'dcos-launch.'.
+                            format('\033[93m'))  # represents the yellow color
+
+            module = 'github.com/dcos/{}?ref={}/{}'.format(
+                'terraform-dcos-enterprise' if self.config['dcos-enterprise'] else 'terraform-dcos',
+                self.config['terraform_dcos_enterprise_version'] if self.config['dcos-enterprise'] else
+                self.config['terraform_dcos_version'], self.config['platform'])
+
+            # Converting our YAML config to the required format. You can find an example of that format in the
+            # Advance YAML Configuration" section here:
+            # https://github.com/mesosphere/terraform-dcos-enterprise/tree/master/aws
+            with open(self.cluster_profile_path, 'w') as file:
+                for k, v in self.config['terraform_config'].items():
+                    file.write(k + ' = ')
+                    if type(k) is dict:
+                        file.write('<<EOF\n{}\nEOF\n'.format(yaml.dump(v)))
+                    else:
+                        file.write('"{}"\n'.format(v))
+            subprocess.run([self.terraform_cmd(), 'init', '-from-module', module], cwd=self.init_dir,
+                           check=True, stderr=subprocess.STDOUT)
+            subprocess.run([self.terraform_cmd(), 'apply', '-auto-approve', '-var-file', self.cluster_profile_path],
+                           cwd=self.init_dir, check=True, stderr=subprocess.STDOUT, env=os.environ)
+        except Exception as e:
+            util.write_json(info_path, self.config)
+            if self.config['auto_rollback']:
+                self.delete()
+            raise e
+        return self.config
+
+    def _install_terraform(self):
+        download_path = os.path.join(self.dcos_launch_root_dir, 'terraform.zip')
+        try:
+            with open(download_path, 'wb') as f:
+                log.info('Downloading...')
+                r = requests.get(self.config['terraform_tarball_url'])
+                for chunk in r.iter_content(1024):
+                    f.write(chunk)
+            with zipfile.ZipFile(download_path, 'r') as tfm_zip:
+                tfm_zip.extractall(self.dcos_launch_root_dir)
+            os.chmod(self.terraform_binary, 0o100)
```

Comment:
I personally prefer to use the abstractions in place of the octal `0o100`. E.G.:
```
import os
import stat

os.chmod(self.terraform_binary, os.stat(self.terraform_binary).st_mode | stat.S_IEXEC)
```
I find this makes it a bit more clear what is happening: you are taking the current permissions and adding exec permissions

Reply:
Alternatively, using the octal and then explaining what this permission is in a simple comment also works fine :)